docs/temporal-explained/workflow-determinism.md

```diff
@@ -0,0 +1,131 @@
+---
+id: workflow-determinism
+title: Workflow Determinism
+sidebar_label: Workflow Determinism
+description: Why workflows must be deterministic, and how you can ensure they are.
+---
+
+# Workflow Determinism
+
+Temporal Workflows must be deterministic. Stated using our terminology:
+
+> For a given [Workflow Type](/docs/concepts/what-is-a-workflow-type), its [Workflow Definition](/docs/concepts/what-is-a-workflow-definition)
+> (implementation) must produce the same sequence of [Commands](/docs/concepts/what-is-a-command) given the same [History](/docs/concepts/what-is-an-event-history)
+
+Less formally, this means a Workflow must always do the same thing given the same inputs.
+
+This requirement is important and necessary because it is how Temporal is able to ensure the state
+of your workflow is not lost if, for example, the [Worker](/docs/concepts/what-is-a-worker) running
+your workflow crashes.
+
+As your Workflow is executed, Commands are generated and appended to the History of the Workflow along
+with other [Events](/docs/concepts/what-is-an-event). If a Workflow's state needs to be re-created
+on a new Worker, the entire History for that [Workflow Execution](/docs/concepts/what-is-a-workflow-execution)
+is sent to the worker and replayed from the beginning. If the same sequence of Commands is emitted
```

Comment:
There's no de-duplication. This description omits the concept of sticky/incremental replay (which isn't really relevant to the concepts here) if that's what you're getting at. I could mention as an aside, it's not always all of the history, technically.

The workflow function is executed (multiple times, actually) during replay, and really we're just checking that the same commands are emitted in the same order, and that every event which applies to some command matches up with the state of that command's state machine.

I could add that last bit of information as well, but I'm worried it just muddies the waters. What really matters is that the same commands are emitted in the same order. If that's true (barring bugs in SDK implementation or incompatibilities between server/sdk), then events can all be applied properly. If it's not true, there's either a problem because you emitted a command where we expected you to emit a different one, or we'll try to apply some event to a nonexistent or wrong type of command.